README.md
# integration-mixpanel

[![CircleCI](https://circleci.com/gh/segment-integrations/integration-mixpanel.svg?style=shield&circle-token=b6b6ffb7aa1d00d99f3fff76dce3a93ca013b5b8)](https://circleci.com/gh/segment-integrations/integration-mixpanel)
  
Mixpanel server-side integration for [Segment](https://segment.com).


HISTORY.md

3.3.2 / 2017-04-17
==================

  * Bump int-tester version to 2.x (#47)

3.3.1 / 2017-03-15
==================

  * Set status code for bad_request errors

3.3.0 / 2017-03-08
==================

  * package: remove "integration-version"

3.1.0 / 2017-01-31
==================

  * Standardize integration (linting, Docker configuration, circle.yml, upgrade
segmentio-integration version, upgrade integration-worker version, etc.)


1.0.22 / 2016-12-12
==================

  * Fix $referrer, $username, and $search_engine 

1.0.21 / 2016-10-25
==================

  * Re-add $name for client side parity 

1.0.20 / 2016-07-06
==================

  * allow bypassing of last seen when tracking revenue

1.0.19 / 2016-05-12
===================

  * Add logic to support consolidatedPageCalls settings for Mixpanel

1.0.18 / 2016-03-02
===================

  * map utm campaign name

1.0.17 / 2016-02-05
===================

  * fix utm bug

1.0.16 / 2016-02-04
===================

  * update fixture timestamps

1.0.15 / 2016-02-04
===================

  * fixing empty callback

1.0.14 / 2016-02-04
===================

  * rejecting timestamps older than 5 years

1.0.13 / 2016-02-03
===================

  * map to way more context props and traits refractor some functions
  * fix multiple page calls issue

1.0.12 / 2016-01-06
===================

  * Merge pull request #12 from segmentio/traits-token
  * Remap traits.token to not conflict with Mixpanels internal keywords.

1.0.11 / 2015-11-20
===================

  * map $app_release

1.0.10 / 2015-09-24
===================

  * $query not a reserved property anymore

1.0.9 / 2015-07-30
==================

  * add support for push notification

1.0.7 / 2015-01-05
==================

  * Fix created/createdAt date formatting issue
  * Update tests, fixtures

1.0.6 / 2014-12-08
==================

 * bump segmentio-integration

1.0.5 / 2014-12-03
==================

  * remove nextTick

1.0.4 / 2014-12-02
==================

 * bump integration proto

1.0.3 / 2014-12-02
==================

 * remove .retries()
 * fix dev deps
 * bump dev deps

1.0.2 / 2014-12-02
==================

 * bump segmentio-integration

1.0.1 / 2014-11-21
==================

 * Bumping segmentio-integration
 * fix build status badge

1.0.0 / 2014-11-14
==================

  * Initial release


test/index.js

var Test = require('segmentio-integration-tester');
var helpers = require('./helpers');
var assert = require('assert');
var time = require('unix-time');
var Mixpanel = require('..');
var uuid = require('uuid');

describe('Mixpanel', function(){
  var mixpanel;
  var settings;
  var test;

  beforeEach(function(){
    settings = {
      apiKey: 'c31daa4f214e00452a9658cadc5ef6de',
      secret: '2692184e7c7a4d9a005b12cf4b9fb22c',
      token: '50912cd33fd82225ab5ae1c563bd5a7e',
      people: true,
      setAllTraitsByDefault: true
    };
    mixpanel = new Mixpanel(settings);
    test = Test(mixpanel, __dirname);
  });

  it('should have correct settings', function(){
    test
      .name('Mixpanel')
      .endpoint('https://api.mixpanel.com')
      .ensure('settings.token')
      .channels(['server']);
  });

  describe('.validate()', function(){
    it('should be invalid when .token is missing', function(){
      delete settings.token;
      test.invalid({}, settings);
    });

    it('should be invalid for old "track" messages without .apiKey', function(){
      delete settings.apiKey;
      test.invalid({
        type: 'track',
        timestamp: new Date('5/10/2014')
      }, settings);
    });

    it('should be valid for new "track" messages without .apiKey', function(){
      delete settings.apiKey;
      test.valid({
        type: 'track',
        timestamp: new Date
      }, settings);
    });

    it('should be valid when all settings are given', function(){
      test.valid({}, settings);
    });

    it('should be invalid for messages with a timestamp older than five years', function(){
      test.invalid({
        type: 'track',
        timestamp: new Date('5/10/2010')
      }, settings);
    });
  });

  describe('.identify()', function(){
    it('should do nothing if `.people` is false', function(done){
      var msg = helpers.identify();
      mixpanel.settings.people = false;
      mixpanel.identify(msg, function(){
        assert.equal(0, arguments.length);
        done();
      });
    });

    it('should send identify correctly', function(done){
      var json = updateFixtureTimestamp(test.fixture('identify-basic'));

      test
        .set(settings)
        .identify(json.input)
        .query({ ip: 0, verbose: 1 })
        .query('data', json.output, decode)
        .expects(200)
        .end(done);
    });

    it('should be able to identify correctly', function(done){
      var msg = helpers.identify();
      mixpanel.identify(msg, done);
    });

    it('should send $set correctly when device token is provided', function(done){
      var json = updateFixtureTimestamp(test.fixture('identify-device-token-set'));

      test
        .set(settings)
        .identify(json.input)
        .request(0)
        .query({ ip: 0, verbose: 1 })
        .query('data', json.output, decode)
        .expects(200)
        .end(done)
    });

    it('should send $union correctly when device token is provided', function(done){
      var json = updateFixtureTimestamp(test.fixture('identify-device-token-union'));

      test
        .set(settings)
        .identify(json.input)
        .request(1)
        .query({ ip: 0, verbose: 1 })
        .query('data', json.output, decode)
        .expects(200)
        .end(done)
    });

    it('should error on invalid request', function(done){
      test
        .set({ apiKey: 'x' })
        .set({ secret: 'x' })
        .set({ token: 'x' })
        .identify({})
        .end(function(err, res) {
          assert.equal(err.status, 400);
          done();
        });
    });


    it('should send identify with context correctly', function(done){
      var json = updateFixtureTimestamp(test.fixture('identify-context'));

      //Test if the context settings have come through properly.
      test
        .set(settings)
        .identify(json.input)
        .query({ ip: 0, verbose: 1 })
        .query('data', json.output, decode)
        .expects(200)
        //What does the query need to do and check here?
        .end(done);
    });

    it('should filter to peopleProperties setting when setAllTraitsByDefault is false', function(done){
      var identify = updateFixtureTimestamp(test.fixture('identify-filter-properties'));
      test
        .set({ setAllTraitsByDefault: false })
        .set({ peopleProperties: [ '$first_name', 'met', '$created', '$email', 'id' ] })
        .identify(identify.input)
        .query({ ip: 0, verbose: 1 })
        .query('data', identify.output, decode)
        .requests(1)
        .expects(200, done);
    })
  });

  describe('.track()', function(){
    it('should send track correctly', function(done){
      var json = test.fixture('track-basic');
      var date = json.input.timestamp = new Date();
      json.output.properties.time = time(date);
      test
        .set(settings)
        .set({ people: false })
        .track(json.input)
        .query({ api_key: settings.apiKey })
        .query({ verbose: '1' })
        .query('data', json.output, decode)
        .requests(1)
        .expects(200, done);
    });

    it('should have proper error status code', function(done){
      var json = test.fixture('track-basic');
      json.input.userId = uuid();
      // sending timestamp far in the future will trigger an error response from their API
      json.input.timestamp = new Date('2123-09-16');
    
      test
        .set(settings)
        .set({ people: false })
        .track(json.input)
        .end(function(err, res){
          console.log(res);
          assert.equal(err.status, 400);
          done();
        });
    });

    it('should send revenue correctly', function(done){
      var json = test.fixture('track-revenue');
      var timestamp = json.input.timestamp = new Date();
      json.output.$append.$transactions.$time = timestamp.toISOString().slice(0,19);

      var suite = test
        .requests(2) // total number of requests
        .set(settings)
        .set({ people: false })
        .track(json.input);

      suite
        .request(1) // second request
        .query({ verbose: '1' })
        .query({ ip: '0' })
        .query('data', json.output, decode)
        .expects(200, done);
    });

    it('should not send last seen with revenue if active flag is false', function(done){
      var json = test.fixture('track-ignore-time-with-revenue');
      var timestamp = json.input.timestamp = new Date();
      json.output.$append.$transactions.$time = timestamp.toISOString().slice(0,19);

      var suite = test
        .requests(2) // total number of requests
        .set(settings)
        .set({ people: false })
        .track(json.input);

      suite
        .request(1) // second request
        .query({ verbose: '1' })
        .query({ ip: '0' })
        .query('data', json.output, decode)
        .expects(200, done);
    });

    it('should send track with context correctly', function(done){
      var json = test.fixture('track-context');
      test
        .set(settings)
        .set({ people: false })
        .track(json.input)
        .query({ api_key: settings.apiKey })
        .query({ verbose: '1' })
        .query('data', json.output, decode)
        .requests(1)
        .expects(200, done);
    });

    it('should send track with only browser context correctly', function(done){
      var json = test.fixture('track-context-browser');
      test
        .set(settings)
        .set({ people: false })
        .track(json.input)
        .query({ api_key: settings.apiKey })
        .query({ verbose: '1' })
        .query('data', json.output, decode)
        .requests(1)
        .expects(200, done);
    });

    it('should send track with referrer correctly', function(done){
      var json = test.fixture('track-referrer');
      test
        .set(settings)
        .set({ people: false })
        .track(json.input)
        .query({ api_key: settings.apiKey })
        .query({ verbose: '1' })
        .query('data', json.output, decode)
        .end(function(err, res){
          if (err) return done(err);
          assert.equal(1, res.length);
          assert.equal(200, res[0].status);
          done();
        });
    });


    // TODO: why are these tests not checking output payload?
    it('should be able to track correctly', function(done){
      mixpanel.track(helpers.track(), done);
    });

    // TODO: why are these tests not checking output payload?
    it('should be able to track a bare call', function(done){
      mixpanel.track(helpers.track.bare(), done);
    });

    it('should increment', function(done){
      var track = helpers.track({ event: 'increment' });
      mixpanel.settings.increments = [track.event()];
      mixpanel.track(track, done);
    })

    it('should be able to track ill-formed traits', function(done){
      mixpanel.track(helpers.track.bare({
        context: {
          traits: 'aaa'
        }
      }), done);
    });
  });

  describe('.alias()', function(){
    it('should be able to alias properly', function(done){
      var json = test.fixture('alias-basic');
      test
        .set(settings)
        .alias(json.input)
        .query('data', json.output, decode)
        .query('api_key', settings.apiKey)
        .query('verbose', '1')
        .query('ip', '0')
        .expects(200)
        .end(done);
    });


    it('should error on invalid request', function(done){
      test
        .set({ apiKey: 'x' })
        .set({ secret: 'x' })
        .set({ token: 'x' })
        .alias({})
        .error('distinct_id, missing or empty', done);
    });
  });

  describe('.page()', function(){
    it('should be able to track all pages', function(done){
      var json = test.fixture('page-all');
      test
        .set(settings)
        .set(json.settings)
        .page(json.input)
        .query('data', json.output, decode)
        .query('api_key', settings.apiKey)
        .query('verbose', '1')
        .expects(200, done);
    });

    it('should be able to track categorized pages', function(done){
      var json = test.fixture('page-categorized');
      test
        .set(settings)
        .set(json.settings)
        .page(json.input)
        .query('data', json.output, decode)
        .query('api_key', settings.apiKey)
        .query('verbose', '1')
        .expects(200, done);
    });

    it('should be able to track categorized pages with consolidated names', function(done){
      var json = test.fixture('page-categorized-consolidated');
      test
        .set(settings)
        .set(json.settings)
        .page(json.input)
        .query('data', json.output, decode)
        .query('api_key', settings.apiKey)
        .query('verbose', '1')
        .expects(200, done);
    });

    it('should be able to track named pages', function(done){
      var json = test.fixture('page-named');
      test
        .set(settings)
        .set(json.settings)
        .page(json.input)
        .query('data', json.output, decode)
        .query('api_key', settings.apiKey)
        .query('verbose', '1')
        .expects(200, done);
    });

    it('should be able to track named pages with the consolidated page name', function(done){
      var json = test.fixture('page-named-consolidated');
      test
        .set(settings)
        .set(json.settings)
        .page(json.input)
        .query('data', json.output, decode)
        .query('api_key', settings.apiKey)
        .query('verbose', '1')
        .expects(200, done);
    });

    it('should not send any requests for disabled pages', function(done){
      var json = test.fixture('page-named');
      test
       .set({
        trackCategorizedPages: false,
        trackNamedPages: false,
        trackAllPages: false
       })
       .page(json.input)
       .requests(0)
       .end(done);
    })
  });

  describe('.screen()', function(){
    it('should be able to track all screens', function(done){
      var json = test.fixture('screen-all');
      test
        .set(settings)
        .set(json.settings)
        .screen(json.input)
        .query('data', json.output, decode)
        .query('api_key', settings.apiKey)
        .query('verbose', '1')
        .expects(200, done);
    });

    it('should be able to track categorized screen', function(done){
      var json = test.fixture('screen-categorized');
      test
        .set(settings)
        .set(json.settings)
        .screen(json.input)
        .query('data', json.output, decode)
        .query('api_key', settings.apiKey)
        .query('verbose', '1')
        .expects(200, done);
    });

    it('should be able to track categorized screens with consolidated names', function(done){
      var json = test.fixture('screen-categorized-consolidated');
      test
        .set(settings)
        .set(json.settings)
        .screen(json.input)
        .query('data', json.output, decode)
        .query('api_key', settings.apiKey)
        .query('verbose', '1')
        .expects(200, done);
    });

    it('should be able to track named screen', function(done){
      var json = test.fixture('screen-named');
      test
        .set(settings)
        .set(json.settings)
        .screen(json.input)
        .query('data', json.output, decode)
        .query('api_key', settings.apiKey)
        .query('verbose', '1')
        .expects(200, done);
    });

    it('should be able to track named screens with the consolidated screen name', function(done){
      var json = test.fixture('screen-named-consolidated');
      test
        .set(settings)
        .set(json.settings)
        .screen(json.input)
        .query('data', json.output, decode)
        .query('api_key', settings.apiKey)
        .query('verbose', '1')
        .expects(200, done);
    });

    it('should not send any requests for disabled screens', function(done){
      var json = test.fixture('screen-named');
      test
       .set({
        trackCategorizedPages: false,
        trackNamedPages: false,
        trackAllPages: false
       })
       .page(json.input)
       .requests(0)
       .end(done);
    })
  });
});

/**
 * Decode base64 and parse json
 */

function decode(data){
  var buf = new Buffer(data, 'base64');
  return JSON.parse(buf.toString());
}

/**
 * Set a fixture's timestamping to the current date.
 *
 * https://mixpanel.com/help/reference/http#storing-user-profiles, $time
 * section: Updates are stored as events in Mixpanel, which are replayed in
 * $time order. If you send an update with the same timestamp as an older
 * update, it won't show up.
 */

function updateFixtureTimestamp(fixture) {
  var date = new Date();
  fixture.input.timestamp = date.toISOString();
  fixture.output.$time = date.getTime();
  return fixture;
}


lib/index.js
'use strict';

/**
 * Module dependencies.
 */

var integration = require('segmentio-integration');
var parse = require('ua-parser-js');
var object = require('obj-case');
var time = require('unix-time');
var extend = require('extend');
var each = require('lodash.foreach');
var reject = require('reject');
var Batch = require('batch');
var tick = setImmediate;
var ms = require('ms');
var is = require('is');
var dates = require('convert-dates');

/**
 * Expose `Mixpanel`
 */

var Mixpanel = module.exports = integration('Mixpanel')
  .endpoint('https://api.mixpanel.com')
  .ensure('settings.token')
  .channels(['server'])
  .retries(2);

/**
 * Mixpanel requires an `.apiKey` on `track` and `screen`
 * if the message is older than 5 days. And
 * won't import anything older than 5 years.
 *
 * https://mixpanel.com/docs/api-documentation/importing-events-older-than-5-days
 * btw, we're not ensuring against this but it looks like if you send a date far into the future
 * Mixpanel will return an error but they do not have this documented anywhere so won't change ensure logic
 */

Mixpanel.ensure(function(msg, settings){
  var age = Date.now() - msg.timestamp();
  if (age > ms('5d')) {
    return this.invalid('message.timestamp() must be within the last five days.');
  }
  if (settings.apiKey) return;
  if (msg.type() !== 'track' && msg.type() !== 'screen') return;
  if (!shouldImport(msg)) return;
  return this.invalid('.apiKey is required if "track" or "screen" message is older than 5 days.');
});

/**
 * Identify the Mixpanel user.
 *
 * https://mixpanel.com/help/reference/http#people-analytics-updates
 *
 * @param {Identify} identify
 * @param {Function} fn
 * @api public
 */

Mixpanel.prototype.identify = function(identify, fn){
  if (!this.settings.people) return tick(fn);

  var batch = new Batch;
  batch.throws(true)
  var self = this;

  var ignoreIp = identify.proxy('context.ignoreIp')  // TODO: remove
    || identify.proxy('context.Mixpanel.ignoreIp');

  var ignoreTime = identify.proxy('context.Mixpanel.ignoreTime')
    || !identify.active();

  // Default querystring and payload data shared across all calls
  var basePayload = {
    $distinct_id: identify.userId() || identify.sessionId(), // the primary id
    $token: this.settings.token,
    $time: identify.timestamp().getTime(),
    $ip: ignoreIp ? 0 : identify.ip() || 0, // use the ip passed in
    $ignore_time: ignoreTime,
    mp_lib: 'Segment: ' + identify.library().name
  };
  var baseQuery = {
    ip: 0,            // pass a flag to ignore the server-ip
    verbose: 1        // make sure that we get a valid response
  };

  // $set must be a separate call than $union
  batch.push(function(fn){
    var payload = extend({ $set: formatTraits(identify) }, basePayload);
    var query = extend({ data: b64encode(payload) }, baseQuery);

    self
      .get('/engage')
      .query(query)
      .end(self._parseResponse(fn));
  });

  // https://mixpanel.com/help/reference/ios-push-notifications
  // Send device token for iOS for push notification support
  var deviceToken = identify.proxy('context.device.token');
  if (deviceToken) {
    batch.push(function(fn){
      var payload = extend({
        $union: {
          $ios_devices: [deviceToken]
        }
      }, basePayload);
      var query = extend({ data: b64encode(payload) }, baseQuery);

      self
        .get('/engage')
        .query(query)
        .end(self._parseResponse(fn));
    });
  }

  batch.end(fn);

};

/**
 * Track a Mixpanel event
 *
 * https://mixpanel.com/help/reference/http#tracking-events
 *
 * TODO: update people's profile when increment is found.
 * see: https://mixpanel.com/help/reference/http#update-operations
 *
 * @param {Track} track
 * @param {Function} fn
 * @api public
 */

Mixpanel.prototype.track = function(track, fn){
  var imported = shouldImport(track)
  var endpoint = imported ? '/import' : '/track';
  var batch = new Batch;
  var self = this;
  var payload = {
    event: track.event(),
    properties: formatProperties(track, this.settings)
  };

  extend(payload.properties, superProperties(track, this.settings));
  var query = {
    verbose: 1,
    data: b64encode(payload),
    api_key: this.settings.apiKey
  };
  // increment
  if (this.settings.people) {
    batch.push(function(fn){
      self.increment(track, fn);
    });
  }

  batch.push(function(done){
    self
    .post(endpoint)
    .set('Content-Length', 0)
    .query(query)
    .end(self._parseResponse(done));
  });

  if (track.revenue()) {
    batch.push(function(done){
      self.revenue(track, done);
    });
  }

  batch.end(fn);
};

/**
 * Track Page / Screen using `msg`.
 *
 * TODO:
 *
 *    In the new Integration proto abstract this away,
 *    doing `if's` in each integration is annoying,
 *    we can automate this even for integrations
 *    that map to page -> track -- this will reduce errors.
 *
 * @param {Track|Screen} msg
 * @param {Function} fn
 */

Mixpanel.prototype.screen =
Mixpanel.prototype.page = function(msg, fn){
  var consolidatedPageCalls = this.settings.consolidatedPageCalls;
  var trackAllPages = this.settings.trackAllPages;
  var trackCategorizedPages = this.settings.trackCategorizedPages;
  var trackNamedPages = this.settings.trackNamedPages;
  var category = msg.category();
  var name = msg.name();
  var fullName = msg.fullName(); // includes both category and name
  var self = this;

  // if they want to send all page/scren events with the default name from facade
  // i.e. "Loaded a Screen", "Loaded a Page"
  if (consolidatedPageCalls) return send(msg.track(), fn);

  // The old behavior
  // track all pages
  if (trackAllPages) return send(msg.track(), fn);

  // track categorized pages
  if (trackCategorizedPages && category){
    if (name) return send(msg.track(fullName), fn);
    return send(msg.track(category), fn);
  }

  // track named pages
  if (trackNamedPages && name) return send(msg.track(name), fn);

  // nothing
  return tick(fn);

  // call via track method with `msg`.
  function send(msg, done){
    self.track(msg, function(err, arr){
      if (err) return done(err);
      done(null, arr[1]);
    });
  }
};

/**
 * Alias a user from one id to the other
 *
 * https://mixpanel.com/help/reference/http#distinct-id-alias
 *
 * @param {Alias} alias
 * @param {Function} fn
 * @api public
 */

Mixpanel.prototype.alias = function(alias, fn){
  var previousId = alias.previousId();
  var userId = alias.userId();
  var payload = {
    event: '$create_alias',
    properties: {
      distinct_id: previousId,
      alias: userId,
      token: this.settings.token
    }
  };

  this
    .post('/track')
    .query({ ip: 0 })
    .query({ verbose: 1 })
    .query({ data: b64encode(payload) })
    .query({ api_key: this.settings.apiKey })
    .set('Content-Length', 0) // mixpanel rejects length-less requests
    .end(this._parseResponse(fn));
};

/**
 * Track a mixpanel revenue call
 *
 * https://mixpanel.com/help/reference/http#tracking-revenue
 *
 * @param {Track} track
 * @param {Function} callback
 * @api private
 */

Mixpanel.prototype.revenue = function(track, fn){
  var ignoreIp = track.proxy('options.ignoreIp');
  if (ignoreIp === undefined) ignoreIp = true;
  var req = this.get('/engage');
  if (ignoreIp) req.query({ ip: 0 });
  req
    .query({ verbose: 1 })
    .query({ data: b64encode(formatRevenue(track, this.settings)) })
    .end(this._parseResponse(fn));
};

/**
 * Increment the given `track` with `settings` and `callback`.
 *
 * Unfortunately Mixpanel doesn't let you specify 2 operations,
 * so we request twice, once for `$add` and once for `$set`.
 *
 * @param {Track} track
 * @param {Function} fn
 * @api private
 */

Mixpanel.prototype.increment = function(track, fn){
  var increments = getIncrements(track, this.settings);
  var batch = new Batch;
  var self = this;

  // ignore
  if (!increments) return tick(fn);
  if (!track.userId()) return tick(fn);

  // send
  batch.push(send('$add'));
  batch.push(send('$set'));
  batch.end(fn);

  // send `type`
  function send(type){
    return function(done){
      var payload = {};
      payload.$distinct_id = track.userId();
      payload.$token = self.settings.token;
      payload.mp_lib = 'Segment.io';
      payload[type] = increments[type];
      var b64 = b64encode(payload);

      return self
        .get('/engage')
        .query({ ip: 0 })
        .query({ verbose: 1 })
        .query({ data: b64 })
        .end(self._parseResponse(done));
    };
  }
};

/**
 * Common function for parsing the response from a mixpanel call.
 *
 * @param {Function} fn
 * @api private
 */

Mixpanel.prototype._parseResponse = function(fn){
  var self = this;
  return this.handle(function(err, res){
    if (err) return fn(err);
    // Mixpanel doesn't send normal error objects with status codes so we have to manually parse their responses
    // for proper error handling by integration-workers
    if (!res.body.status) {
      var message = res.body.error;
      var newError = new Error(message);
      newError.status = message.match(/api_key/) ? 401 : 400;
      fn(newError);
    } else {
      fn(null, res);
    }
  });
};

/**
 * Add user super properties to the track.
 *
 * @param {Track} track
 * @return {Object}
 * @api private
 */

function superProperties(track, settings){
  var identify = track.identify();
  var traits = formatTraits(identify) || {};
  var properties = {};

  if (!is.object(traits)) return properties;

  Object.keys(traits).forEach(function(trait){
    var val = traits[trait];
    // an early version of the integrations prefixed traits incorrectly
    // the setting preserves backwards compat.
    if (settings.legacySuperProperties && trait.charAt(0) !== '$') {
      trait = '$' + trait;
    }
    properties[trait] = val;
  });

  return properties;
}

/**
 * A map of Segment traits to special Mixpanel user properties.
 *
 * https://mixpanel.com/help/reference/http#people-special-properties
 */

var traitAliases = {
  created: '$created',
  createdAt: '$created',
  email: '$email',
  firstName: '$first_name',
  lastName: '$last_name',
  lastSeen: '$last_seen', // This should be removed since docs say not to set this manually or can't be set
  name: '$name',
  username: '$username',
  phone: '$phone',
  // Token is reserved by Mixpanel, so remap it.
  token: 'trait_token'
};

/**
 * Format the traits from the identify
 *
 * @param {Identify} identify
 * @return {Object}
 * @api private
 */

function formatTraits(identify){
  // Get traits, renaming any special Mixpanel properties from their Segment names
  var traits = identify.traits(traitAliases) || {};
  var userAgent = identify.userAgent();

  // Delete any Segment trait names; they've now been renamed to Mixpanel names
  each(traitAliases, function(_, key) {
    object.del(traits, key, { normalizer: function(path) {
      return path.replace(/[^A-Za-z0-9\.$]+/g, '').toLowerCase();
    }});
  });

  if (userAgent) extend(traits, formatUserAgent(userAgent));

  // Format timestamp
  // https://mixpanel.com/help/reference/http#people-special-properties, `$created` section
  if (traits.$created) {
    traits.$created = formatDate(traits.$created);
  }

  // Map semantic mobile context properties
  extend(traits, formatMobileSpecific(identify));

  stringifyValues(traits);
  return traits;
}

/**
 * Format the mixpanel specific properties.
 *
 * https://github.com/mixpanel/mixpanel-android/blob/34eb2205882cae137597dfb16f9c13545de32ca5/src/main/java/com/mixpanel/android/mpmetrics/AnalyticsMessages.java#L403
 *
 * @param {Track} track
 * @param {Object} settings
 * @return {Object}
 */

function formatProperties(track, settings){
  var properties = track.properties() || {};
  var identify = track.identify();
  var userAgent = track.userAgent();
  var campaign = track.proxy('context.campaign') || undefined;
  var app = track.proxy('context.app') || {};
  var semanticProps = {
    $app_release: app.build,
    $app_version: app.version,
    $current_url: track.proxy('context.page.url'),
    $device: track.proxy('context.device.name'),
    distinct_id: track.userId() || track.sessionId(),
    ip: track.ip(),
    mp_lib: 'Segment: ' + track.library().name,
    $referrer: track.referrer(),
    $search_engine: track.proxy('properties.searchEngine'),
    time: time(track.timestamp()),
    token: settings.token,
    $username: track.proxy('properties.username')
  };

  // Remove possible duplicate properties.
  // Note that this will object.del will strip any special characters
  // before the lookup
  object.del(properties, 'username');
  object.del(properties, 'searchEngine');
  object.del(properties, 'referrer');

  extend(properties, semanticProps);

  // Add the name tag
  properties.mp_name_tag = identify.name()
    || identify.email()
    || identify.userId()
    || identify.sessionId();

  if (userAgent) extend(properties, formatUserAgent(userAgent));

  // Map mobile specific special props
  extend(properties, formatMobileSpecific(track));

  // Map UTM params
  if (campaign) {
    properties.utm_source = campaign.source;
    properties.utm_medium = campaign.medium;
    properties.utm_term = campaign.term;
    properties.utm_content = campaign.content;
    properties.utm_campaign = campaign.name;
  }

  // Strip null/undefined values
  properties = reject(properties);
  stringifyValues(properties);

  return properties;
}

/**
 * Format userAgent properties
 *
 * https://mixpanel.com/help/questions/articles/what-properties-do-mixpanels-libraries-store-by-default
 *
 * @param {string} data
 * @return {Object}
 * @api private
 */

function formatUserAgent(data){
  var ret = {};
  var parsed = parse(data);
  var browser = parsed.browser;
  var os = parsed.os;

  if (browser) {
    ret.$browser = browser.name;
    ret.$browser_version = browser.version;
  }

  return ret;
}

/**
 * Format mobile specific properties
 * https://mixpanel.com/help/questions/articles/what-properties-do-mixpanels-libraries-store-by-default
 *
 * @param {msg} msg
 * @return {Object}
 * @api private
 */

function formatMobileSpecific(track) {
  var device = track.proxy('context.device') || {};
  var app = track.proxy('context.app') || {};
  var os = track.proxy('context.os') || {};
  var network = track.proxy('context.network') || {};
  var screen = track.proxy('context.screen') || {};
  var ret = {};

  ret.$carrier = network.carrier;
  ret.$manufacturer = device.manufacturer;
  ret.$model = device.model;
  ret.$os = os.name;
  ret.$os_version = os.version;
  ret.$screen_height = screen.height;
  ret.$screen_width = screen.width;
  ret.$wifi = network.wifi
  // ret.$brand = device.model; what should this be


  switch (device.type) {
    case 'iOS':
      ret.$ios_ifa = device.advertisingId;
      ret.$ios_device_model = device.model;
      ret.$ios_app_release = app.build;
      ret.$ios_app_version = app.version;
      ret.$ios_version = os.version;
      break;
    case 'android':
      ret.$screen_dpi = screen.density;
      ret.$bluetooth_enabled = network.bluetooth;
      ret.$has_telephone = network.cellular;
      ret.$android_version_code = app.version;
      ret.$android_app_version = app.version;
      ret.$android_os = os.name;
      ret.$android_os_version = os.version;
      ret.$android_model = device.model;
      ret.$android_manufacturer = device.manufacturer;
      // ret.$android_brand = device.manufacturer; ?? what should this be
      break;
  }
  return ret;
}

/**
 * Create a revenue track call
 *
 * https://mixpanel.com/help/reference/http#tracking-revenue
 *
 * @param {Track} track
 * @param {Object} settings
 * @return {Object}
 * @api private
 */

function formatRevenue(track, settings){
  var ret = {
    $distinct_id: track.userId() || track.sessionId(),
    $token: settings.token,
    $ip: track.ip(),
    $append: {
      $transactions: {
        $time: formatDate(track.timestamp()),
        $amount: track.revenue()
      }
    }
  };
  // don't flag as last seen by default
  var ignoreTime = !track.active();
  if (ignoreTime) ret.$ignore_time = ignoreTime;
  return ret;
}

/**
 * Formats a date for Mixpanel's API, takes the first part of the iso string
 *
 * https://mixpanel.com/help/reference/http#dates-in-updates
 *
 * @param {Mixed} date
 * @return {String}
 * @api private
 */

function formatDate(date){
  date = new Date(date);
  if (isNaN(date.getTime())) return;
  return date.toISOString().slice(0,19);
}

/**
 * Get increments.
 *
 * @param {Track} track
 * @param {Object} settings
 * @return {Object}
 * @api private
 */

function getIncrements(track, settings){
  var inc = lowercase(settings.increments || []);
  var event = track.event();
  if (!~inc.indexOf(event.toLowerCase())) return;
  var ret = { $set: {}, $add: {} };
  ret.$set['Last ' + event] = formatDate(track.timestamp());
  ret.$add[event] = 1;
  return ret;
}

/**
 * Mixpanel uses different endpoints for historical import.
 *
 * https://mixpanel.com/docs/api-documentation/importing-events-older-than-31-days
 *
 * @param {Facade} message
 * @return {Boolean}
 * @api private
 */

function shouldImport(message){
  var timestamp = message.timestamp() || new Date();
  return (Date.now() - timestamp.getTime()) > ms('5d');
}

/**
 * Base64 encode the payload
 *
 * @param {Object} payload
 * @return {String}
 * @api private
 */

function b64encode(payload){
  return new Buffer(JSON.stringify(payload)).toString('base64');
}

/**
 * Stringify the nested values for an object.
 *
 * @param {Object} obj
 * @return {Object}
 * @api private
 */

function stringifyValues(obj){
  if (!is.object(obj)) return obj;

  Object.keys(obj).forEach(function(key){
    var val = obj[key];
    if (is.object(val)) obj[key] = JSON.stringify(val);
  });

  return obj;
}

/**
 * Lowercase the given `arr`.
 *
 * @param {Array} arr
 * @return {Array}
 * @api private
 */

function lowercase(arr){
  var ret = [];

  for (var i = 0; i < arr.length; ++i) {
    ret.push(String(arr[i]).toLowerCase());
  }

  return ret;
}
